objects/src/accounts/delta/vault.rs

```diff
@@ -1,236 +1,488 @@
+    /// Returns an iterator over all keys that have the provided action.
+    fn filter_by_action(
+        &self,
+        action: NonFungibleDeltaAction,
+    ) -> impl Iterator<Item = Digest> + '_ {
+        self.0
+            .iter()
+            .filter(move |&(_, cur_action)| cur_action == &action)
+            .map(|(key, _)| *key)
+    }
+
+    /// Checks whether this vault delta is valid.
+    ///
+    /// # Errors
+    /// Returns an error if the number of updates is greater than [u16::MAX].
+    fn validate(&self) -> Result<(), AccountDeltaError> {
+        if self.0.len() > u16::MAX as usize {
+            return Err(AccountDeltaError::TooManyNonFungibleAssets {
+                actual: self.0.len(),
+                max: u16::MAX as usize,
+            });
+        }
+
+        Ok(())
     }
 }
 
-impl Serializable for AccountVaultDelta {
+impl Serializable for NonFungibleAssetDelta {
     fn write_into<W: ByteWriter>(&self, target: &mut W) {
-        assert!(self.added_assets.len() <= u16::MAX as usize, "too many added assets");
-        target.write_u16(self.added_assets.len() as u16);
-        target.write_many(self.added_assets.iter());
+        let added: Vec<_> = self.filter_by_action(NonFungibleDeltaAction::Add).collect();
+        let removed: Vec<_> = self.filter_by_action(NonFungibleDeltaAction::Remove).collect();
 
-        assert!(self.removed_assets.len() <= u16::MAX as usize, "too many removed assets");
-        target.write_u16(self.removed_assets.len() as u16);
-        target.write_many(self.removed_assets.iter());
+        target.write_usize(added.len());
+        target.write_many(added.iter());
+
+        target.write_usize(removed.len());
+        target.write_many(removed.iter());
     }
 }
 
-impl Deserializable for AccountVaultDelta {
+impl Deserializable for NonFungibleAssetDelta {
     fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
-        // deserialize and validate added assets
-        let num_added_assets = source.read_u16()? as usize;
-        let mut added_assets: Vec<Asset> = Vec::with_capacity(num_added_assets);
-        for _ in 0..num_added_assets {
-            let asset = Asset::read_from(source)?;
-            if added_assets.iter().any(|a| a.is_same(&asset)) {
-                return Err(DeserializationError::InvalidValue(
-                    "asset added more than once".to_string(),
-                ));
-            }
+        let num_added = source.read_usize()?;
+        let added = source.read_many::<Digest>(num_added)?;
+
+        let num_removed = source.read_usize()?;
+        let removed = source.read_many::<Digest>(num_removed)?;
+
+        let map = added
+            .into_iter()
+            .map(|key| (key, NonFungibleDeltaAction::Add))
+            .chain(removed.into_iter().map(|key| (key, NonFungibleDeltaAction::Remove)))
+            .collect();
+
+        Self::new(map).map_err(|err| DeserializationError::InvalidValue(err.to_string()))
```

Comment:
I think we can optimize this a bit to avoid extra vector allocations by using loops.